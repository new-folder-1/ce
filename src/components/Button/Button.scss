.Button {
    height: 30px;
    
    border-radius: 2px;
    font-weight: bold;
    color: #fff;

    &[disabled] {
        opacity: 0.5;
    }

    &_theme_default {
        border: 1px solid #fff;
        background-color: #255dd3;
    }
    &_theme_action {
        border: none;
        background-color: #f57c00;
    }
}
