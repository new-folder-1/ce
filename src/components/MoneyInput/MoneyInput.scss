.MoneyInput {
    height: 32px;
    font-size: 20px;
    background-color: transparent;
    color: #fff;
    text-align: right;
    border: none;
    outline: none;
    width: 100%;

    &::placeholder {
        color: #fff;
        opacity: 0.3;
    }

    &[disabled] {
        opacity: 0.6;
    }
}
