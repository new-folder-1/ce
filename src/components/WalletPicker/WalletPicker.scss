.WalletPicker {
    color: #fff;

    &-Currency {
        font-weight: bold;
        font-size: 24px;
        line-height: 32px;
    }

    &-Balance, &-Rate {
        font-size: 13px;
    }

    &-Row {
        display: flex;
        justify-content: space-between;
        align-items: center;

        margin-bottom: 12px;
    }

    &-Input {
        height: 32px;
        font-size: 20px;
        background-color: transparent;
        color: #fff;
        text-align: right;
    }
}
