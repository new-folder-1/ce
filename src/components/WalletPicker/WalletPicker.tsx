import * as React from 'react';

import { Wallet, Currency } from "../../types";

interface WalletPickerProps {
    type: WalletType;

    current: Wallet;
    currentIndex: number;

    amount: number;

    rate: number;
    exchangeCurrency: Currency;

    walletsAmount: number;
    onBallanceChange: (type: WalletType, value: string) => void;
    onWalletChange: (type: WalletType, walletIndex: number) => void;
}

export type WalletType = 'from' | 'to';

export const isNumber = (value: string) => /^[0-9]*[.]{0,1}[0-9]{0,2}$/.test(value);

export const WalletPicker = ({
    type,
    onBallanceChange,
    onWalletChange,
    current,
    currentIndex,
    walletsAmount,
    rate,
    exchangeCurrency,
    amount
}: WalletPickerProps) => {
    const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        console.log(isNumber(e.target.value));
        if (!isNumber(e.target.value)) {
            return;
        }
        onBallanceChange(type, e.target.value);
    }
    const onPrevClick = () => {
        onWalletChange(type, currentIndex === 0 ? 0 : currentIndex - 1);
    }
    const onNextClick = () => {
        onWalletChange(type, currentIndex === walletsAmount - 1 ? currentIndex : currentIndex + 1);
    }
    console.log(type, amount);
    return (
        <div>
            <h4>{current.currency}</h4>
            <span>You have: {current.currency} {current.amount}</span>
            <input onChange={onChange} value={+amount.toFixed(5)} />
            {type === 'to' && (
                <div>
                    {current.currency}1 = {exchangeCurrency}{Number(rate.toFixed(5))}
                </div>
            )}
            <div>
                <button onClick={onPrevClick}>Prev</button>
                <button onClick={onNextClick}>Next</button>
            </div>
        </div>
    );
};
