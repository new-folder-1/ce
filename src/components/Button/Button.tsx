import * as React from 'react';

import './Button.scss';

interface ButtonProps {
    text: string;
    disabled?: boolean;
    onClick?: () => void;
    className?: string;
    theme?: ButtonTheme;
}

type ButtonTheme = 'default' | 'action';

export const Button = React.memo(({ text, theme, className, ...rest }: ButtonProps) => { 
    const bClassName = `Button Button_theme_${theme || 'default'} ${className || ''}`;
    return (
        <button
            className={bClassName}
            {...rest}
        >
            {text}
        </button>
    );
});
