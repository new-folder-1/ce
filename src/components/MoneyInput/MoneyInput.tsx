import * as React from 'react';
import NumberFormat, { NumberFormatValues } from "react-number-format";

import './MoneyInput.scss';

export interface MoneyInputProps {
    prefix: '+' | '-' | '';
    amount?: number;
    autofocus?: boolean;
    onChange: (value: number) => void;
}

export const MoneyInput =({
    prefix, amount, autofocus, onChange
} : MoneyInputProps) => {
    const onNumberInput = React.useCallback((value: NumberFormatValues) => {
        const newValue = Math.abs(value.floatValue);
        if (newValue === amount) {
            return;
        }
    }, [onChange, amount]);

    return (
        <NumberFormat
            className="MoneyInput"
            prefix={`${prefix} `}
            allowNegative={false}
            value={amount}
            decimalScale={2}
            decimalSeparator="."
            onValueChange={onNumberInput}
            placeholder="0.00"
            autoFocus={autofocus}
            type="tel"
        />
    );
};
