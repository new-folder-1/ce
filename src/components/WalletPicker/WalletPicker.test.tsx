import React from 'react';
import { shallow } from 'enzyme';

import { WalletPicker, WalletPickerProps } from './WalletPicker';
import { Button } from '../Button/Button';
import { MoneyInput } from '../MoneyInput/MoneyInput';

const baseProps: WalletPickerProps = {
    type: 'from',
    prev: {
        currency: 'USD',
        amount: 200,
        id: '1'
    },
    current: {
        currency: 'GBP',
        amount: 300,
        id: '2'
    },
    next: {
        currency: 'EUR',
        amount: 100.20,
        id: '3'
    },
    onWalletChange: jest.fn(),
    onAmountChange: jest.fn(),
    rate: 0.8992,
    exchangeCurrency: 'USD',
    amount: 20.50
};

describe('WalletPicker', () => {
    it('snapshot renders', () => {
        const component = shallow(<WalletPicker {...baseProps} />);
        expect(component).toMatchSnapshot();
    });

    it('prev is not rendered', () => {
        const props: WalletPickerProps = {
            ...baseProps,
            prev: undefined
        };
        const component = shallow(<WalletPicker {...props} />);
        expect(component).toMatchSnapshot();
    });

    it('next is not rendered', () => {
        const props: WalletPickerProps = {
            ...baseProps,
            next: undefined
        };
        const component = shallow(<WalletPicker {...props} />);
        expect(component).toMatchSnapshot();
    });

    it('calls onWalletChange when buttons are clicked', () => {
        const component = shallow(<WalletPicker {...baseProps} />);
        const buttons = component
            .find('.WalletPicker-WalletChanger')
            .find(Button);
        
        buttons.at(0).prop('onClick')();
        expect(baseProps.onWalletChange).toBeCalledWith('prev');

        buttons.at(1).prop('onClick')();
        expect(baseProps.onWalletChange).toBeCalledWith('next');
    });

    it('calls onAmountChange when input changes', () => {
        const component = shallow(<WalletPicker {...baseProps} />);
        const inputs = component.find(MoneyInput);

        inputs.at(0).prop('onChange')(200);
        expect(baseProps.onAmountChange).toBeCalledWith('from', 200);
    });
});
