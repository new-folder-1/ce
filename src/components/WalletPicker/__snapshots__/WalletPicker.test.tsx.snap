// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletPicker next is not rendered 1`] = `
<section
  className="WalletPicker"
>
  <div
    className="WalletPicker-Row"
  >
    <header
      className="WalletPicker-Currency"
    >
      GBP
    </header>
    <Memo()
      amount={20.5}
      autofocus={true}
      onChange={[Function]}
      prefix="-"
    />
  </div>
  <div
    className="WalletPicker-Row"
  >
    <span
      className="WalletPicker-Balance"
    >
      You have: 
      <Memo()
        amount={300}
        currency="GBP"
      />
    </span>
    <span
      className="WalletPicker-Rate"
    >
      <Memo()
        amount={1}
        currency="GBP"
      />
       = 
      <Memo()
        amount={0.8992}
        currency="USD"
      />
    </span>
  </div>
  <div
    className="WalletPicker-Row WalletPicker-WalletChanger"
  >
    <Memo()
      onClick={[Function]}
      text="← USD"
    />
    <span>
       
    </span>
  </div>
</section>
`;

exports[`WalletPicker prev is not rendered 1`] = `
<section
  className="WalletPicker"
>
  <div
    className="WalletPicker-Row"
  >
    <header
      className="WalletPicker-Currency"
    >
      GBP
    </header>
    <Memo()
      amount={20.5}
      autofocus={true}
      onChange={[Function]}
      prefix="-"
    />
  </div>
  <div
    className="WalletPicker-Row"
  >
    <span
      className="WalletPicker-Balance"
    >
      You have: 
      <Memo()
        amount={300}
        currency="GBP"
      />
    </span>
    <span
      className="WalletPicker-Rate"
    >
      <Memo()
        amount={1}
        currency="GBP"
      />
       = 
      <Memo()
        amount={0.8992}
        currency="USD"
      />
    </span>
  </div>
  <div
    className="WalletPicker-Row WalletPicker-WalletChanger"
  >
    <span>
       
    </span>
    <Memo()
      onClick={[Function]}
      text="EUR →"
    />
  </div>
</section>
`;

exports[`WalletPicker snapshot renders 1`] = `
<section
  className="WalletPicker"
>
  <div
    className="WalletPicker-Row"
  >
    <header
      className="WalletPicker-Currency"
    >
      GBP
    </header>
    <Memo()
      amount={20.5}
      autofocus={true}
      onChange={[Function]}
      prefix="-"
    />
  </div>
  <div
    className="WalletPicker-Row"
  >
    <span
      className="WalletPicker-Balance"
    >
      You have: 
      <Memo()
        amount={300}
        currency="GBP"
      />
    </span>
    <span
      className="WalletPicker-Rate"
    >
      <Memo()
        amount={1}
        currency="GBP"
      />
       = 
      <Memo()
        amount={0.8992}
        currency="USD"
      />
    </span>
  </div>
  <div
    className="WalletPicker-Row WalletPicker-WalletChanger"
  >
    <Memo()
      onClick={[Function]}
      text="← USD"
    />
    <span>
       
    </span>
    <Memo()
      onClick={[Function]}
      text="EUR →"
    />
  </div>
</section>
`;
